service/rpccommon/server.go

```diff
@@ -0,0 +1,343 @@
+var typeOfError = reflect.TypeOf((*error)(nil)).Elem()
+
+// Is this an exported - upper case - name?
+func isExported(name string) bool {
+	rune, _ := utf8.DecodeRuneInString(name)
+	return unicode.IsUpper(rune)
+}
+
+// Is this type exported or a builtin?
+func isExportedOrBuiltinType(t reflect.Type) bool {
+	for t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+	// PkgPath will be non-empty even for an exported type,
+	// so we need to check the type name as well.
+	return isExported(t.Name()) || t.PkgPath() == ""
+}
+
+// Fills methods map with the methods of receiver that should be made
+// available through the RPC interface.
+// These are all the public methods of rcvr that have one of those
+// two signatures:
+// func (rcvr ReceiverType) Method(in InputType, out *ReplyType) error
+// func (rcvr ReceiverType) Method(in InputType, cb service.RPCCallback
+func suitableMethods(rcvr interface{}, methods map[string]*methodType) {
+	typ := reflect.TypeOf(rcvr)
+	rcvrv := reflect.ValueOf(rcvr)
+	sname := reflect.Indirect(rcvrv).Type().Name()
+	if sname == "" {
+		log.Printf("rpc.Register: no service name for type %s", typ)
+		return
+	}
+	for m := 0; m < typ.NumMethod(); m++ {
+		method := typ.Method(m)
+		mname := method.Name
+		mtype := method.Type
+		// method must be exported
+		if method.PkgPath != "" {
+			continue
+		}
+		// Method needs three ins: (receive, *args, *reply) or (receiver, *args, *RPCCallback)
+		if mtype.NumIn() != 3 {
+			log.Println("method", mname, "has wrong number of ins:", mtype.NumIn())
+			continue
+		}
+		// First arg need not be a pointer.
+		argType := mtype.In(1)
+		if !isExportedOrBuiltinType(argType) {
+			log.Println(mname, "argument type not exported:", argType)
+			continue
+		}
+
+		replyType := mtype.In(2)
+		synchronous := replyType.String() != "service.RPCCallback"
+
+		if synchronous {
+			// Second arg must be a pointer.
+			if replyType.Kind() != reflect.Ptr {
+				log.Println("method", mname, "reply type not a pointer:", replyType)
+				continue
+			}
+			// Reply type must be exported.
+			if !isExportedOrBuiltinType(replyType) {
+				log.Println("method", mname, "reply type not exported:", replyType)
+				continue
+			}
+
+			// Method needs one out.
+			if mtype.NumOut() != 1 {
+				log.Println("method", mname, "has wrong number of outs:", mtype.NumOut())
+				continue
+			}
+			// The return type of the method must be error.
+			if returnType := mtype.Out(0); returnType != typeOfError {
+				log.Println("method", mname, "returns", returnType.String(), "not error")
+				continue
+			}
+		} else {
+			// Method needs zero outs.
+			if mtype.NumOut() != 0 {
+				log.Println("method", mname, "has wrong number of outs:", mtype.NumOut())
+				continue
+			}
+		}
+		methods[sname+"."+mname] = &methodType{method: method, ArgType: argType, ReplyType: replyType, Synchronous: synchronous, Rcvr: rcvrv}
+	}
+	return
+}
+
+func (s *ServerImpl) serveCodec(conn io.ReadWriteCloser) {
```

Comment:
nit: This isn't a generic codec, we are specifically running the json rpc codec. Maybe rename to `serveJSONCodec` to be more specific.